portfolio-project-matching-app/backend/dao.js

```diff
@@ -425,38 +475,64 @@ const readDocIdsByCriteria = async (coll, field, criteria) => {
     return querySnap.docs.map(doc => doc.id);
 }
 
-const readApplicationByApplicationId = async (id) => {
+const readAllDocs = async (coll) => {
     /*
-    DESCRIPTION:    retrieves application document by application document id
+    DESCRIPTION:    retrieves all Firebase documents for specified collection
 
-    INPUT:          id (string) : document ID of desired application
+    INPUT:          coll (string): collection to get documents from
 
-    RETURN:         Application object
+    RETURN:         array of Firebase documents with ID included
     */
-    // get snapshot for invalid input handling
-    const docSnap = await getDocSnapshotById('applications', id);
-    if (!docSnap.exists()) {
-        console.log(`Invalid id: 'applications' does not have document id '${id}'`);
-        return -1;
-    }
-    // handle case of valid field
-    else {
-        // build "base" Application object
-        const app = Application.fromDocSnapshot(docSnap.id, docSnap);
-        // populate project, user, and owner properties with applicable objects
-        [app.project, app.user, app.owner] = await Promise.all([
-            getShallowProjectById(app.projectId),
-            getShallowUserById(app.userId),
-            getShallowUserById(app.ownerId)
-        ])
-        // return array of Application objects to calling function
-        return app;
-    }
+    // get collection snapshot
+    const collSnap = await getCollectionSnapshot(coll);
+    // handle case of empty collection snapshot
+    if (collSnap.empty) return [];
+    // handle case of non-empty collection
+    return collSnap.docs.map((doc) => {
+        return {...doc.data(), id: doc.id};
+    });
+}
+
+const readAssociationObjectsByType = async (assocColl, assocField, assocId, coll, field, deep = false) => {
+    /*
+    DESCRIPTION:    returns an array of associations based on arguments passed.
+                    Note, this function is not for the feint of heart :P
+
+    INPUT:          assocColl (string): association collection that stores
+                    the relationship
+
+                    assocField (string): field pertaining to the document you
+                    want to get associations for. E.g., if the user wants
+                    projects by user => assocField = user_id
+
+                    assocId (string): document ID of the document user desires
+                    associations for
+
+                    coll (string): collection where the associations live.
+                    If the user wants projects by user => coll = 'projects'
+
+                    field (string): field in assocColl pertaining to the
+                    document ID of the associations
+
+    RETURN:         array of objects
+    */   
```

Comment:
This function is a bit confusing, but cuts down on about three million different `getTechnologiesByUser()`, `getUsersByProject()`, etc. Unfortunately, it requires an obnoxious number of parameters. I actually tried cutting down on the parameters, but felt like it didn't really help all that much. Just made it more difficult to understand what the heck you were actually querying.

Note: this is mostly used internally by other `dao` functions. The only place it shows up in the project is `myProjects`, because we need **deep** project associations by user.